### PR TITLE
Migrating host side code to OpenSSL 3.0

### PR DESCRIPTION
**Pull request by justanotherminh**
All these functions have been migrated to OpenSSL 3.0

File: ~/openenclave/common/crypto/openssl/ec.c
_private_key_write_pem_callback()
oe_ec_public_key_from_coordinates()
oe_ec_generate_key_pair_from_private()
oe_ec_valid_raw_private_key()

File: openenclave/common/crypto/openssl/rsa.c
_private_key_write_pem_callback()
_get_public_key_get_modulus_or_exponent()
_public_key_equal()

File: openenclave/host/crypto/openssl/rsa.c
oe_rsa_get_public_key_from_private()
oe_rsa_public_key_from_modulus()

File: ~/openenclave/common/crypto/openssl/sha.c
oe_sha256_init()
oe_sha256_update()
oe_sha256_final()

File: ~/openenclave/host/crypto/openssl/init.c
_initialize()

File: ~/openenclave/common/crypto/openssl/hmac.c
oe_hmac_sha256_init()
oe_hmac_sha256_update()
oe_hmac_sha256_final()

OpenSSL functions that are deprecated in the 3.0 release: 
EVP_PKEY_get1_EC_KEY()
PEM_write_bio_ECPrivateKey()
EVP_PKEY_get1_EC_KEY()
EC_KEY_get0_group()
EC_KEY_get0_public_key()
EC_KEY_new()
EC_KEY_set_group()
EC_POINT_set_affine_coordinates_GFp()
EC_KEY_set_public_key()
EVP_PKEY_assign_EC_KEY()
EC_KEY_new_by_curve_name()
EC_KEY_set_asn1_flag()
EVP_PKEY_set1_EC_KEY()
EC_GROUP_clear_free()
PEM_write_bio_RSAPrivateKey()
RSA_get0_key()
RSA_set0_key()
EVP_PKEY_get1_RSA()
SHA256_Init()
SHA256_Update()
SHA256_Final()
HMAC_CTX_init()
HMAC_CTX_new()
HMAC_Init_ex()
HMAC_Update()
HMAC_Final()

### CODE REVIEW

**Comment by radhikaj** [comment]
Add information on what functions are deprecated

**Comment by radhikaj** [comment]
bors try

**Comment by justanotherminh** [comment]
bors try

**Comment by mingweishih** [comment]
bors try

File: common/crypto/openssl/ec.c [hunk 1]
```diff
@@ -320,6 +384,96 @@ oe_result_t oe_ec_generate_key_pair_from_private(
...
+            point,
+            POINT_CONVERSION_UNCOMPRESSED,
+            buffer,
+            keysize,
+            NULL) > 1024)
+        OE_RAISE(OE_CRYPTO_ERROR);
+    // Generate EVP_PKEY for private key
+    if (!(public_pkey = EVP_PKEY_new()))
+        OE_RAISE(OE_CRYPTO_ERROR);
+    if (!(EVP_PKEY_set_bn_param(private_pkey, "priv", priv)))
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:444]: Is `priv` the required name or it can be arbitrary?

File: common/crypto/openssl/ec.c [hunk 2]
```diff
@@ -320,6 +384,96 @@ oe_result_t oe_ec_generate_key_pair_from_private(
...
+        OE_RAISE(OE_CRYPTO_ERROR);
+    // Generate EVP_PKEY for public key
+    if (!(ctx = EVP_PKEY_CTX_new_from_name(NULL, "EC", NULL)))
+        OE_RAISE(OE_CRYPTO_ERROR);
+    EVP_PKEY_fromdata_init(ctx);
+    params[0] = OSSL_PARAM_construct_utf8_string(
+        "group", (char*)OBJ_nid2sn(_get_nid(curve)), 0);
+    params[1] = OSSL_PARAM_construct_octet_string("pub", buffer, keysize);
+    params[2] = params[2] = OSSL_PARAM_construct_end();
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:453]: Is `params[2] = params[2] = ...` intentional?

File: common/crypto/openssl/ec.c [hunk 3]
```diff
@@ -320,6 +384,96 @@ oe_result_t oe_ec_generate_key_pair_from_private(
...
+        OE_RAISE(OE_CRYPTO_ERROR);
+    if (EC_POINT_point2oct(
+            group,
+            point,
+            POINT_CONVERSION_UNCOMPRESSED,
+            buffer,
+            keysize,
+            NULL) > 1024)
+        OE_RAISE(OE_CRYPTO_ERROR);
+    // Generate EVP_PKEY for private key
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:441]: The comment seems wrong

File: common/crypto/openssl/ec.c (hunk 1 above)
Comment (mingweishih) [reply common/crypto/openssl/ec.c:444]: Doesn't the `private_pkey` require initialization via `EVP_PKEY_new()`?

File: common/crypto/openssl/ec.c [hunk 4]
```diff
@@ -320,6 +384,96 @@ oe_result_t oe_ec_generate_key_pair_from_private(
...
+{
+    oe_result_t result = OE_UNEXPECTED;
+    int openssl_result;
+    EC_GROUP* group = NULL;
+    EVP_PKEY_CTX* ctx = NULL;
+    EVP_PKEY* public_pkey = NULL;
+    EVP_PKEY* private_pkey = NULL;
+    BIGNUM* priv = NULL;
+    EC_POINT* point = NULL;
+    uint8_t buffer[1024];
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:403]: How 1024 is determined?

File: common/crypto/openssl/ec.c [hunk 5]
```diff
@@ -320,6 +384,96 @@ oe_result_t oe_ec_generate_key_pair_from_private(
...
+    size_t keysize;
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:404]: Initialize to zero

File: common/crypto/openssl/ec.c [hunk 6]
```diff
@@ -320,6 +384,96 @@ oe_result_t oe_ec_generate_key_pair_from_private(
...
+            buffer,
+            keysize,
+            NULL) > 1024)
+        OE_RAISE(OE_CRYPTO_ERROR);
+    // Generate EVP_PKEY for private key
+    if (!(public_pkey = EVP_PKEY_new()))
+        OE_RAISE(OE_CRYPTO_ERROR);
+    if (!(EVP_PKEY_set_bn_param(private_pkey, "priv", priv)))
+        OE_RAISE(OE_CRYPTO_ERROR);
+    // Generate EVP_PKEY for public key
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:446]: The comment refers to line 447 - 455?

File: common/crypto/openssl/ec.c [hunk 7]
```diff
@@ -440,6 +595,93 @@ oe_result_t oe_ec_public_key_from_coordinates(
...
+    if (!public_key || !x_data || !x_size || x_size > OE_INT_MAX || !y_data ||
+        !y_size || y_size > OE_INT_MAX)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Get the NID for this curve type */
+    if ((nid = _get_nid(ec_type)) == NID_undef)
+        OE_RAISE(OE_FAILURE);
+
+    /* Create the public EC key */
+    {
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:633]: The bracket can be removed

File: common/crypto/openssl/ec.c [hunk 8]
```diff
@@ -440,6 +595,93 @@ oe_result_t oe_ec_public_key_from_coordinates(
...
+        if (!(param_bld = OSSL_PARAM_BLD_new()))
+            OE_RAISE(OE_CRYPTO_ERROR);
+
+        if (!OSSL_PARAM_BLD_push_utf8_string(
+                param_bld, "group", OBJ_nid2sn(nid), 0))
+            OE_RAISE(OE_CRYPTO_ERROR);
+        unsigned char pub_data[1 + x_size + y_size];
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:640]: Use malloc instead for the variable-size buffer

File: common/crypto/openssl/ec.c [hunk 9]
```diff
@@ -440,6 +595,93 @@ oe_result_t oe_ec_public_key_from_coordinates(
...
+        if (OSSL_PARAM_BLD_push_octet_string(
+                param_bld, "pub", pub_data, sizeof(pub_data)))
+            OE_RAISE(OE_CRYPTO_ERROR);
+
+        params = OSSL_PARAM_BLD_to_param(param_bld);
+
+        ctx = EVP_PKEY_CTX_new_from_name(NULL, "EC", NULL);
+        if (ctx == NULL || params == NULL || EVP_PKEY_fromdata_init(ctx) <= 0 ||
+            EVP_PKEY_fromdata(ctx, &pkey, EVP_PKEY_PUBLIC_KEY, params) <= 0)
+
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:653]: Extra empty line?

File: common/crypto/openssl/ec.c [hunk 10]
```diff
@@ -440,6 +595,93 @@ oe_result_t oe_ec_public_key_from_coordinates(
...
+            OE_RAISE(OE_CRYPTO_ERROR);
+    }
+
+    /* Create the PKEY public key wrapper */
+    {
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:658]: Remove the bracket

File: common/crypto/openssl/ec.c [hunk 11]
```diff
@@ -560,7 +802,7 @@ bool oe_ec_valid_raw_private_key(
     if (bn != NULL)
         BN_clear_free(bn);
     if (group != NULL)
-        EC_GROUP_clear_free(group);
+        EC_GROUP_free(group);
```
Comment (mingweishih) [inline common/crypto/openssl/ec.c:805]: Is this a bug fix (for both OpenSSL 1.1.1 and 3)?

File: common/crypto/openssl/hmac.c [hunk 12]
```diff
@@ -26,7 +34,16 @@ static void _free_hmac_ctx(HMAC_CTX* ctx)
     HMAC_CTX_free(ctx);
 #endif
 }
+#else
+static void _free_hmac_ctx(EVP_MAC_CTX* ctx)
+{
+    if (!ctx)
+        return;
+    EVP_MAC_CTX_free(ctx);
```
Comment (mingweishih) [inline common/crypto/openssl/hmac.c:42]: Add an empty line above

File: common/crypto/openssl/hmac.c [hunk 13]
```diff
@@ -67,6 +84,45 @@ oe_result_t oe_hmac_sha256_init(
...
+    ctx = EVP_MAC_CTX_new(mac);
+    if (ctx == NULL)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+    OSSL_PARAM params[3];
+    params[0] = OSSL_PARAM_construct_octet_string(
+        OSSL_MAC_PARAM_KEY, (const void*)key, keysize);
+    params[1] = OSSL_PARAM_construct_utf8_string(
+        OSSL_MAC_PARAM_DIGEST, OSSL_DIGEST_NAME_SHA2_256, 0);
+    params[2] = OSSL_PARAM_construct_end();
+    openssl_result = EVP_MAC_init(ctx, NULL, 0, params);
```
Comment (mingweishih) [inline common/crypto/openssl/hmac.c:111]: Use `if (!EVP_MAC_init(...` and remove the openssl_result

File: common/crypto/openssl/hmac.c [hunk 14]
```diff
@@ -67,6 +84,45 @@ oe_result_t oe_hmac_sha256_init(
...
+
+    if (!context || !key || keysize > OE_INT_MAX)
+        OE_RAISE(OE_INVALID_PARAMETER);
+    EVP_MAC* mac = EVP_MAC_fetch(NULL, OSSL_MAC_NAME_HMAC, NULL);
+    if (!mac)
+        OE_RAISE(OE_CRYPTO_ERROR);
+    ctx = EVP_MAC_CTX_new(mac);
+    if (ctx == NULL)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+    OSSL_PARAM params[3];
```
Comment (mingweishih) [inline common/crypto/openssl/hmac.c:105]: Move the declaration to the beginning of the function.

File: common/crypto/openssl/hmac.c [hunk 15]
```diff
@@ -67,6 +84,45 @@ oe_result_t oe_hmac_sha256_init(
...
+    const uint8_t* key,
+    size_t keysize)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    EVP_MAC_CTX* ctx = NULL;
+    int openssl_result;
+
+    if (!context || !key || keysize > OE_INT_MAX)
+        OE_RAISE(OE_INVALID_PARAMETER);
+    EVP_MAC* mac = EVP_MAC_fetch(NULL, OSSL_MAC_NAME_HMAC, NULL);
```
Comment (mingweishih) [inline common/crypto/openssl/hmac.c:99]: Move the declaration to the beginning of the function.

File: common/crypto/openssl/rsa.c [hunk 16]
```diff
@@ -52,7 +53,34 @@ static oe_result_t _private_key_write_pem_callback(BIO* bio, EVP_PKEY* pkey)
 
     return result;
 }
+#else
+static oe_result_t _private_key_write_pem_callback(BIO* bio, EVP_PKEY* pkey)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    unsigned char* buffer = NULL;
+    size_t bytes_written;
```
Comment (mingweishih) [inline common/crypto/openssl/rsa.c:61]:
```suggestion
    size_t bytes_written = 0;
```

File: common/crypto/openssl/rsa.c [hunk 17]
```diff
@@ -52,7 +53,34 @@ static oe_result_t _private_key_write_pem_callback(BIO* bio, EVP_PKEY* pkey)
...
+        OE_RAISE(OE_CRYPTO_ERROR);
+    BIO_write(bio, buffer, (int)bytes_written);
+
+    result = OE_OK;
+
+done:
+
+    if (ctx)
+        OSSL_ENCODER_CTX_free(ctx);
+
```
Comment (mingweishih) [inline common/crypto/openssl/rsa.c:78]: Should buffer be freed as well?

File: common/crypto/openssl/rsa.c [hunk 18]
```diff
@@ -52,7 +53,34 @@ static oe_result_t _private_key_write_pem_callback(BIO* bio, EVP_PKEY* pkey)
...
+    oe_result_t result = OE_UNEXPECTED;
+    unsigned char* buffer = NULL;
+    size_t bytes_written;
+    OSSL_ENCODER_CTX* ctx = OSSL_ENCODER_CTX_new_for_pkey(
+        pkey, EVP_PKEY_KEYPAIR, "PEM", NULL, NULL);
+    if (!ctx)
+        OE_RAISE(OE_CRYPTO_ERROR);
+    if (!OSSL_ENCODER_to_data(ctx, &buffer, &bytes_written))
+        OE_RAISE(OE_CRYPTO_ERROR);
+    if (bytes_written <= 0)
```
Comment (mingweishih) [inline common/crypto/openssl/rsa.c:68]: Might check `buffer` as well

File: common/crypto/openssl/rsa.c [hunk 19]
```diff
@@ -121,6 +149,79 @@ static oe_result_t _get_public_key_get_modulus_or_exponent(
...
+static oe_result_t _get_public_key_get_modulus_or_exponent(
+    const oe_public_key_t* public_key,
+    uint8_t* buffer,
+    size_t* buffer_size,
+    bool get_modulus)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    size_t required_size;
+    const BIGNUM* bn;
+    // RSA* rsa = NULL;
```
Comment (mingweishih) [inline common/crypto/openssl/rsa.c:162]: remove commented code

File: common/crypto/openssl/rsa.c [hunk 20]
```diff
@@ -121,6 +149,79 @@ static oe_result_t _get_public_key_get_modulus_or_exponent(
...
 }
+#else
+static oe_result_t _get_public_key_get_modulus_or_exponent(
+    const oe_public_key_t* public_key,
+    uint8_t* buffer,
+    size_t* buffer_size,
+    bool get_modulus)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    size_t required_size;
```
Comment (mingweishih) [inline common/crypto/openssl/rsa.c:160]:
```suggestion
    size_t required_size = 0;
```

File: common/crypto/openssl/rsa.c [hunk 21]
```diff
@@ -121,6 +149,79 @@ static oe_result_t _get_public_key_get_modulus_or_exponent(
...
+    const BIGNUM* bn;
```
Comment (mingweishih) [inline common/crypto/openssl/rsa.c:161]:
```suggestion
    const BIGNUM* bn = NULL;
```

File: common/crypto/openssl/rsa.c [hunk 22]
```diff
@@ -121,6 +149,79 @@ static oe_result_t _get_public_key_get_modulus_or_exponent(
...
+    /* Check for invalid parameters */
+    if (!public_key || !buffer_size)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* If buffer is null, then buffer_size must be zero */
+    if (!buffer && *buffer_size != 0)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Get RSA key */
+    // if (!(rsa = EVP_PKEY_get1_RSA(public_key->pkey)))
```
Comment (mingweishih) [inline common/crypto/openssl/rsa.c:173]: Remove unused code

File: common/crypto/openssl/rsa.c [hunk 23]
```diff
@@ -121,6 +149,79 @@ static oe_result_t _get_public_key_get_modulus_or_exponent(
...
+    //     OE_RAISE(OE_CRYPTO_ERROR);
+
+    /* Select modulus or exponent */
+    const BIGNUM* e = NULL;
```
Comment (mingweishih) [inline common/crypto/openssl/rsa.c:177]: Move declarations to the beginning of the function

File: common/crypto/openssl/rsa.c [hunk 24]
```diff
@@ -121,6 +149,79 @@ static oe_result_t _get_public_key_get_modulus_or_exponent(
...
+    const BIGNUM* n = NULL;
+    // RSA_get0_key(rsa, &n, &e, NULL);
```
Comment (mingweishih) [inline common/crypto/openssl/rsa.c:179]: remove

**Review by mingweishih — COMMENTED** [review]
Just finished a first pass. Will take another look after the revisions.

**Comment by mingweishih** [comment]
bors try

File: common/crypto/openssl/ec.c (hunk 4 above)
Comment (justanotherminh) [reply common/crypto/openssl/ec.c:403]: Changed to OPENSSL_malloc(keysize)

File: common/crypto/openssl/ec.c (hunk 7 above)
Comment (justanotherminh) [reply common/crypto/openssl/ec.c:649]: Removing this bracket causes the compilation to fail due to unreachable goto label

File: common/crypto/openssl/ec.c (hunk 11 above)
Comment (justanotherminh) [reply common/crypto/openssl/ec.c:821]: EC_group_clear_free is no longer available in 3, and apparently EC_group_free is supposed to cover the same functionality.

**Comment by justanotherminh** [comment]
bors try

**Comment by CyanDevs** [comment]
Hi Minh, a couple of initial points:

1. For your commits you need to have DCO signoff. Please see https://github.com/apps/dco for detailed instructions
2. Could you please rebase this PR to master or resolve the conflicts?
3. Is your openssl submodule suppose to be pointing to their `openssl-3.0` as it refers to `openssl/crypto/bn/rsa_sup_mul.c` which doesn't exist in the current submodule reference commit `ad4910fad2`
4. There are multiple rules generating `3rdparty/openssl/build/crypto/aes/aesni-x86_64.s` in `3rdparty/openssl/CMakeLists.txt` could you please fix this?

Thanks

**Pull request closed without merging**